idea-plugin/src/main/kotlin/arrow/meta/ide/MetaExtensionsRegistrar.kt

```diff
@@ -0,0 +1,44 @@
+package arrow.meta.ide
+
+import com.intellij.ide.ApplicationInitializedListener
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.project.impl.ProjectLifecycleListener
+import org.jetbrains.kotlin.config.CompilerConfiguration
+
+class MetaExtensionsRegistrar : ApplicationInitializedListener {
+  companion object {
+    private val LOG = Logger.getInstance("#arrow.metaRegistrar")
+  }
+
+  override fun componentsInitialized() {
+    val app = ApplicationManager.getApplication()
+
+    // register application-level extensions
```

Comment:
How can do we register application-level extensions
Most of our ide dsl is based on Application-level extensions

Reply:
@i-walker The current dsl is already setting up application-level extensions, e.g. using [this code](https://github.com/arrow-kt/arrow-meta/blob/725d3754b4ca0899c5bbd517e28c072da058fd77/idea-plugin/src/main/kotlin/arrow/meta/ide/internal/registry/IdeInternalRegistry.kt#L130). That's fine for the current set of supported extensions.
The issue is that this code is called too often.

Part of the problem is that the registration above is triggered by `InternalRegistry.registerMetaComponents(project, configuration)`. This takes a project and therefore can't be executed in application scope. 
This way IDE application extensions, IDE project extensions (none atm) and compiler plugins are all registered within a project scope. 

As far as I can tell the lifecycle of IDE extensions is not related to the lifecycle of compiler plugins. At least application-level IDE extensions should not be forced into the lifecycle of compiler plugins, which requires a project.

Currently IDE extensions are kept here:
```
open class IdeMetaPlugin : MetaPlugin(), IdeInternalRegistry, IdeSyntax {
  @ExperimentalContracts
  override fun intercept(ctx: CompilerContext): List<Plugin> =
    super.intercept(ctx) +
      initialIdeSetUp +
      higherKindsIdePlugin +
      typeclassesIdePlugin +
      comprehensionsIdePlugin +
      opticsIdePlugin +
      nothingIdePlugin
}
```

It might be possible to store the IDE plugins in a separate list and offer a method `IdeMetaPlugin.registerApplicationExtensions()` or similar which just iterates on that list. `MetaPluginRegistrar` should then call this method in `componentsInitialized()`.

Reply:
Yes, this sounds legitimate. We might want to add @raulraja to this, because that would mean we either have to abstract away which services are instantiated at what time or we might have to add another cycle to the `intercept` methode

Reply:
I think it sounds like a great idea. There would be a bit of a refactor involved but if @jansorg is able to propose a composition to satisfy the same amount of flexibility, it could be worth a PR in itself